flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/sources/TableSourceUtil.scala

```diff
@@ -139,13 +138,15 @@ object TableSourceUtil {
     *
     * @param tableSource The [[TableSource]] for which the Calcite schema is generated.
     * @param selectedFields The indices of all selected fields. None, if all fields are selected.
+    * @param watermarkSpec The watermark information (passed in from DDL)
     * @param streaming Flag to determine whether the schema of a stream or batch table is created.
     * @param typeFactory The type factory to create the schema.
     * @return The Calcite schema for the selected fields of the given [[TableSource]].
     */
   def getRelDataType(
       tableSource: TableSource[_],
       selectedFields: Option[Array[Int]],
+      watermarkSpec: Option[WatermarkSpec],
```

Comment:
don't think this should be taken care by **TableSource**Util. TableSource only responsible for reading data out of underlying storage, watermark is generated after that.